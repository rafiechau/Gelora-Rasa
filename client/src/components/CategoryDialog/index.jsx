import InputTextField from '@components/InputTextField';
import CloseSharpIcon from '@mui/icons-material/CloseSharp';
import { Box, Button, Dialog, DialogContent, DialogTitle, IconButton, useMediaQuery } from '@mui/material';
import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { FormattedMessage, injectIntl } from 'react-intl';
import { useDispatch } from 'react-redux';
import { actionCreateCategories, actionUpdateCategoryById } from '@pages/Dashboard/Categories/actions';
import { useTheme } from '@emotion/react';

export const CategoryDialog = ({ open, onClose, intl: { formatMessage }, currentCategory }) => {
  const dispatch = useDispatch();
  const isEditMode = !!currentCategory;

  const theme = useTheme();
  const fullScreen = useMediaQuery(theme.breakpoints.down('sm'));

  const {
    handleSubmit,
    register,
    setValue,
    formState: { errors },
  } = useForm();

  useEffect(() => {
    if (isEditMode && currentCategory) {
      setValue('categoryName', currentCategory.name);
    }
  }, [currentCategory, setValue, isEditMode]);

  const onSubmit = (data) => {
    if (isEditMode) {
      dispatch(actionUpdateCategoryById(currentCategory.id, data));
    } else {
      dispatch(actionCreateCategories(data));
    }
    onClose();
  };

  return (
    <Dialog open={open} onClose={onClose} fullWidth maxWidth="sm" fullScreen={fullScreen} data-testid="category-dialog">
      <DialogTitle>
        <FormattedMessage id={isEditMode ? 'app_header_edit_category' : 'app_header_create_category'} />
        <IconButton onClick={onClose} style={{ position: 'absolute', right: 8, top: 8 }} data-testid="close-button">
          <CloseSharpIcon />
        </IconButton>
      </DialogTitle>
      <DialogContent>
        <Box
          component="form"
          noValidate
          onSubmit={handleSubmit(onSubmit)}
          sx={{ mt: 2, display: 'flex', flexDirection: 'column', gap: 2 }}
        >
          <InputTextField
            input={{
              name: 'categoryName',
              required: formatMessage({ id: 'app_category_name_require_message' }),
              type: 'text',
              label: formatMessage({ id: 'app_category_name' }),
            }}
            register={register}
            errors={errors}
          />
          <Button type="submit" fullWidth variant="contained" sx={{ mt: 3, mb: 2 }} data-testid="submit-button">
            {isEditMode ? 'app_btn_submit' : 'app_btn_edit'}
          </Button>
        </Box>
      </DialogContent>
    </Dialog>
  );
};

CategoryDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  intl: PropTypes.object,
  currentCategory: PropTypes.object,
};

export default injectIntl(CategoryDialog);
