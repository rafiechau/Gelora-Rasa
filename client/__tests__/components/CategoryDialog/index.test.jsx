import { render, fireEvent, screen } from '@testing-library/react';
import CategoryDialog from '@components/CategoryDialog';
import { ThemeProvider, createTheme } from '@mui/material';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

jest.mock('react-intl', () => ({
  injectIntl: jest.fn((Component) => (props) => <Component {...props} intl={{ formatMessage: (msg) => msg.id }} />),
  FormattedMessage: jest.fn(({ id }) => <span>{id}</span>),
}));

const mockClose = jest.fn();
const mockCurrentCategory = {
  id: 1,
  categoryName: 'Sample Category',
};

const theme = createTheme();

describe('CategoryDialog', () => {
  beforeEach(() => {
    render(
      <ThemeProvider theme={theme}>
        <CategoryDialog open onClose={mockClose} currentCategory={mockCurrentCategory} />
      </ThemeProvider>
    );
  });

  test('Should render correctly', () => {
    expect(screen.getByTestId('category-dialog')).toBeInTheDocument();
  });

  test('Should call onClose when close button is clicked', () => {
    const closeButton = screen.getByTestId('close-button');
    fireEvent.click(closeButton);
    expect(mockClose).toHaveBeenCalled();
  });

  test('Should render category name in edit mode', () => {
    const input = screen.getByLabelText('app_category_name');
    expect(input.value).toBe(mockCurrentCategory.categoryName);
  });

  test('Should match with snapshot', () => {
    const dialog = screen.getByTestId('category-dialog');
    expect(dialog).toMatchSnapshot();
  });
});
