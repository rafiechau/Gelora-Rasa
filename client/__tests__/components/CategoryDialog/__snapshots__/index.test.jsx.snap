// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryDialog Should match with snapshot 1`] = `
<div
  class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
  data-testid="category-dialog"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    tabindex="-1"
  >
    <div
      aria-labelledby=":r3:"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-tlc64q-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <h2
        class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
        id=":r3:"
      >
        <span>
          app_header_edit_category
        </span>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="close-button"
          style="position: absolute; right: 8px; top: 8px;"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CloseSharpIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </h2>
      <div
        class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
      >
        <form
          class="MuiBox-root css-h4tqr2"
          novalidate=""
        >
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <label
              class="labelInput"
              for="categoryName"
            >
              app_category_name
              <input
                class="input undefined"
                id="categoryName"
                name="categoryName"
                type="text"
                value=""
              />
            </label>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
            data-testid="submit-button"
            tabindex="0"
            type="submit"
          >
            app_btn_submit
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </div>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
